### PR TITLE
remove only use of Apache BVal @NotEmpty annotation (in InteropIdentifierRequest/Response)

### PR DESCRIPTION
**Pull request by vorburger**
I'm not sure if this even worked?  There's no test coverage for it.

This is required to be able to (easily) switch from openjpa-all to openjpa in upcoming #607

The alternative would be to add an explicit dependency to BVal; but that seems unnecessary, and avoiding it will likely reduce other dependency issues.

### CODE REVIEW

**Comment by vorburger** [comment]
@conradsp wondering if you'd be willing to review this one, and merge if OK for you?

PS: Bval = https://bval.apache.org

**Comment by conradsp** [comment]
@marta-jankovics - just FYI, we are removing the notempty annotation as indicated above. This should not affect functionality of interoperability layer. The NotNull checks should take care of any edge cases.

@vorburger - could you please add a NotNull annotation on the InteropIdentifierRequestData constructor for the accountId field: 

`public InteropIdentifierRequestData(@NotNull InteropIdentifierType idType, @NotNull String idValue, String subIdOrType, String accountId) {	    public InteropIdentifierRequestData(@NotNull InteropIdentifierType idType, @NotNull String idValue, String subIdOrType, String accountId) 
`

should become:

`public InteropIdentifierRequestData(@NotNull InteropIdentifierType idType, @NotNull String idValue, String subIdOrType, String accountId) {	    public InteropIdentifierRequestData(@NotNull InteropIdentifierType idType, @NotNull String idValue, String subIdOrType, @NotNull String accountId) 
`

After that, it looks good to me and I can merge (or you can go ahead and merge).

**Comment by marta-jankovics** [comment]
> @marta-jankovics - just FYI, we are removing the notempty annotation as indicated above. This should not affect functionality of interoperability layer. The NotNull checks should take care of any edge cases.
> 
> @vorburger - could you please add a NotNull annotation on the InteropIdentifierRequestData constructor for the accountId field:
> 
> `public InteropIdentifierRequestData(@NotNull InteropIdentifierType idType, @NotNull String idValue, String subIdOrType, String accountId) {	public InteropIdentifierRequestData(@NotNull InteropIdentifierType idType, @NotNull String idValue, String subIdOrType, String accountId) `
> 
> should become:
> 
> `public InteropIdentifierRequestData(@NotNull InteropIdentifierType idType, @NotNull String idValue, String subIdOrType, String accountId) {	public InteropIdentifierRequestData(@NotNull InteropIdentifierType idType, @NotNull String idValue, String subIdOrType, @NotNull String accountId) `
> 
> After that, it looks good to me and I can merge (or you can go ahead and merge).

@conradsp Hi Conrad, thank you for your notice! Without checking the concrete code fragment you are going to modify:
@NotEmpty
"The annotated element must not be {@code null} nor empty."
@NotNull 
"The annotated element must not be {@code null}"
so to me it seems that if you replace @NotEmpty with @NotNull you would affect the functionality.

**Comment by vorburger** [comment]
>  so to me it seems that if you replace @notempty with @NotNull you would affect the functionality.

@marta-jankovics yes, I agree, this affects functionality in the sense that it no longer checks for empty String in a few of the InteropIdentifierRequest/Response fields.

The point here, which I should have made much more clear in the initial comment proposing this, is the following: This [Apache BVal](https://bval.apache.org) `@NotEmpty` annotation is used ONLY in InteropIdentifierRequest/Response, and nowhere else in all of Fineract. (I've updated the title to more accurately reflect this.)

This use blocks #607, which blocks #601, which would help to clean up with seems to be a classpath mess.

So I'm proposing that we simplify things by removing our dependency on BVal, understanding that we loose a (minor?) convenience (?) use of `@NotEmpty` annotation. I doubt that was actually ever intended to be available at all - it's, likely, more of a side effect of the (wrong) use of `openjpa-all` (which is a "shaded" JAR which contains far to much). 

I'm also not entirely sure if this  `@NotEmpty` annotation check even actually worked? At least [the OpenJPA documentation makes no mention of Apache BVal, or even any annotation based validation](http://openjpa.apache.org/builds/2.4.3/apache-openjpa/docs/manual.html).

Long story short, I'm suggesting this (very minor) change despite it (possibly, not even sure) causing a minor regression in checking for (unlikely?) empty fields in one particular request.

The alternative would be to add a real integration test for `@NotEmpty`, and if that currently works, add an explicit dependency to BVal, or some other artifact JAR with these validation annotations. To me, both seem unnecessary, and avoiding it will likely reduce other dependency issues (in #601 I'd like to try to clean up what, to me, currently seems to be a bit of a mess in terms of Fineracts depenencies).

**Comment by vorburger** [comment]
Oh and BTW I just stumbled upon this in `fineract-provider/build.gradle` :

    exclude module: 'bval-core'
    exclude module: 'org.apache.bval.bundle'
    exclude module: 'bval-jsr303'

So in the past someone actually tried to actively prevent the use of Apache BVal at some point - but because of `openjpa-all`, it sneaked (back?) in; this, with #607, will clean that up.

**Comment by vorburger** [comment]
Erm... further work on #607 has now revealed that we actually very much do need a Bean Validation provider, such as Apache BVal (or Hibernate Validator; but we currently use BVal, through `openjpa-all`); without one, Spring start-up fails saying: `javax.validation.NoProviderFoundException: Unable to create a Configuration, because no Bean Validation provider could be found. Add a provider like Hibernate Validator (RI) to your classpath.`

I'm therefore closing this PR, as I'll likely be adding an explicit dependency to BVal over in #607 after all.

Sorry for the noise! :smiling_imp:

**Comment by vorburger** [comment]
Nah, re-opening this after all... what we really should do here, to avoid future confusion, is just replace `org.apache.bval.constraints.NotEmpty` with `javax.validation.constraints.NotEmpty` ... just done - OK for you all to merge this like?  (PS: I'll, probably make Apache BVal an `implementation` instead of `compile` scoped dependency in #607; see [FINERACT-777](https://issues.apache.org/jira/browse/FINERACT-777).

**Comment by vorburger** [comment]
GitHub someone doesn't let me re-open this PR anymore, so #617.

**Comment by marta-jankovics** [comment]
> GitHub someone doesn't let me re-open this PR anymore, so #617.

Hi @vorburger I agree with you, javax.validation.constraints.NotEmpty should be fine.
